### PR TITLE
Create Sandcastle Standalone Mode

### PR DESCRIPTION
**Pull request by OmarShehata**
This is attempt 2 to create a standalone Sandcastle model. This is the previous attempt https://github.com/AnalyticalGraphicsInc/cesium/pull/7097.

@hpinkos @mramato what do you think?

This is a refactor in that it works exactly the same way from the user's prespective. Any Sandcastle URL like:

```
https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/#c=ZYw7C8Iw...
```

Can be run in "standalone" mode by adding `standalone.html`: 

```
https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/standalone.html#c=ZYw7C8Iw...
```

This will work regardless if the demo is loaded from the gallery, from a custom base64 string, or from a GitHub gist (and whatever else Sandcastle does that I am not aware of).

It also opens this mode instead of the blob URL when clicking "Open in new window" button. This fixes that button (which currently will not load ion imagery) as well as creating a way to view Sandcastle examples on mobile. 

And best of all, it's a completely client side solution! 

The only catch is it's kind of a hack. The standalone mode is actually just the same Sandcastle page with everything hidden and a few HTML edits. Ideally we'd have a responsive view but we don't, and this is the next best thing.

What is nice about this is that there's nothing to maintain. It's the exact same behavior/code. 

### Can standalone.html just be index.html with a query parameter?

Yeah but then it won't hide elements until JavaScript runs which will cause a very visible jitter and look ugly.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @OmarShehata!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/standlone-sandcastle/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by OmarShehata** [comment]
> With that in mind, does standalone.html need so much structure? Could it be essentially an empty shell, like bucket.html, and inject the demo into itself, rather than injecting it into a child iframe?

That's exactly the approach I started with, and I agree that would be much nicer. I had to start stripping down `CesumSandcastle.js` into `SandcastleStandalone.js` since it relies a lot on the existence of these divs. It started to feel like I was rebuilding Sandcastle and I was running into issues with it not initiating the demo because it was waiting on some div. So I went back and tried this even simpler approach. 

Part of it is my unfamiliarity with how Sandcastle dynamically loads and runs demos. If all we need to run the demo is to put the bucket in a div on the page, what's the iframe currently there for?

**Comment by OmarShehata** [comment]
@emackey I was halfway through implementing this new architecture when I realized I was getting an error when adding the script tag before the HTML (because it was executing immediately). Which meant I didn't really need to do the `activateBucketScripts` at all as long as I wait for the added HTML to load (like in my Sandcastle CSS PR). 

This works in Chrome, Edge, Firefox and IE11. This means we can just:

1. Something quite similar to bucket-requirejs.html loads first (page is blank).
2. Get the HTML from the query parameter and add it.
3. Once HTML is loaded, add the script tag. 

And that's it. Am I missing anything here?

I'll clean this up and push it in the next few hours.

**Comment by emackey** [comment]
Sounds great!

**Comment by OmarShehata** [comment]
Ok, I've finished refactoring this. It's a lot simpler now, and definitely feels a lot faster I think without all the baggage of dojo/dijit in the new window!

I abstracted away a couple functions in `Sandcastle-helpers.js`. One thing I had trouble with was getting it to use the right version in the unbuilt/built versions. It does it now by using `require` for the unbuilt and just a normal script tag for the built, because doing:

```
<script type="text/javascript" src="../../Source/Cesium.js"></script>
```

Did not work (because that file needs require).

I had a bit of a revelation while working on this: I was getting the old Sandcastle CSS issue https://github.com/AnalyticalGraphicsInc/cesium/issues/5265 very consistently when adding the HTML first and then the JavaScript. Using the MutationObserver didn't actually work because waiting for the DOM element to be added wasn't enough, it seemed like it still had to load. I realized that just listening for:

```
window.addEventListener("load", function(event) {
```

Works perfectly. Which makes me wonder if that would just be a better way to do it in `CesiumSandcastle.js` as opposed to using the MutationObserver. 

But for now, this should be clean and working. Let me know how this looks @emackey .

File: Apps/Sandcastle/standalone.html [hunk 1]
```diff
@@ -0,0 +1,69 @@
+﻿<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no">
+    <title>Cesium Demo</title>
+    <script type="text/javascript" src="Sandcastle-header.js"></script>
+    <script type="text/javascript" src="ThirdParty/pako.min.js"></script>
+    <script type="text/javascript" src="../../../ThirdParty/requirejs-2.1.20/require.js"></script>
```
Comment (emackey) [inline Apps/Sandcastle/standalone.html]: You're one level too deep here, remove one `../`

(I run Cesium from IIS in a sub-folder, and it catches stuff like this)

**Comment by OmarShehata** [comment]
Ok, so I was only able to recreate this on Firefox. Turns out the `load` event just gets fired too early (before the page really finishes loading). The problem isn't really that the HTML element doesn't fully load before the script runs (before the widgets are created by the script anyway), it's that the CSS imported in that HTML doesn't finish getting applied before the script runs.

There's no robust way to check for if the CSS has finished loading. Best I could find was this trick of declaring some CSS, and checking for when it has been applied in a loop. This now works on all the browsers I could find. Can you give it another try @emackey ?

**Comment by OmarShehata** [comment]
Ok, @emackey , it's not the ideal solution (which I think would be just flattening everything in one folder and using the `gallery-index.js` to label the development ones, but that seems like too big of a refactor at the moment), but I got it to work by just adding a query parameter that alters the paths to have it to work. 

I'm not very familiar with running the development examples, but it should be working now for development and normal examples.

**Comment by mramato** [comment]
Did we try IE11? (I'm on Linux right now so I can't test)

File: Apps/Sandcastle/standalone.html [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    var code;
+    //Handle case where demo is in a sub-directory by modifying
+    //the HTML to add a base href.
+    var urlParams = new URLSearchParams(window.location.search);
+
+    function update() {
+        var width = getComputedStyle(loaderElement).getPropertyValue('width');
+        var done = false;
+        if (width === '12px') {
+            frameDelay --;
```
Comment (mramato) [inline Apps/Sandcastle/standalone.html]: spacing

**Comment by mramato** [comment]
Why is `gallery` a query parameter instead of part of the hash?  You normally don't use query parameters for things that don't involve the server.

File: Apps/Sandcastle/CesiumSandcastle.js [hunk 3]
```diff
@@ -110,6 +111,8 @@ require({
     if (!defined(window.Cesium)) {
         window.Cesium = Cesium;
     }
+    // Used by Sandcastle-helpers.js
```
Comment (mramato) [inline Apps/Sandcastle/CesiumSandcastle.js]: Why doesn't sandcastle-helpers use AMD syntax and include pako itself?

**Comment by OmarShehata** [comment]
My bad. The built version broke because I forgot to update the build script with the latest commit. It should all work now. Here's what I tested:

* [x] Works on unbuilt, for regular and development in Chrome.
* [x] Works on built, for regular and development in Chrome. 
* [x] Works on Firefox for above cases. 
* [ ] Works on IE11 for above cases.
* [x] Works when hosted from a static server via (`python -m http.server`).

The only thing that isn't working is in IE11, _some_ of the examples open `standalone.html` without any hash, even though the current data is passed to `window.open`. I wonder if this is just some URL limit it's hitting. Everything else that I tested there should be working. 

I also had to do a workaround for IE11, everything worked except for some reason the base href wasn't getting applied to the `templates/bucket.css` being imported. This was an issue even when I hardcoded the `<base>` tag at the top of the document. 

Does it matter that in the built version I do `/Build/CesiumUnminified/Cesium.js` instead of `../../CesiumUnminified/Cesium.js` ? I think the issue there is that the document head `<base>` is added after that loads, so while `../../CesiumUnminified/Cesium.js` is the correct URL, subsequent requests are incorrect. I'm not sure if there's an easy way around that.

Other Todo:

* [ ] Import Pako with AMD syntax in sandcastle helpers
* [ ] I haven't tried having the base href be relative, but does it matter?

**Comment by OmarShehata** [comment]
Ok, good news. I added a "compile bucket.css" step (https://github.com/AnalyticalGraphicsInc/cesium/pull/7250/commits/f90f1c7472bd65ff6a1226a5508f9f0c2e7b3ace) in the build process, which correctly fixes the CSS race conditions both in standlone and in the regular Sandcastle. It also greatly simplifies `standalone.html` because we don't have to do add a dummy HTML element and poll for the CSS to be added.

I tested all the combinations of built, unbuilt, for regular and development examples, for Chrome, Edge, IE11 and Firefox. The only issue with IE11 is when the URL is too long.

I spent too long trying to get `Sandcastle-helpers.js` to use AMD require syntax. The problem I ran into was that when built, `requirejs` isn't included, so I would need to separately build and include `Sandcastle-helpers`. However this is not done for `CesiumSandcastle.js`. Apparently the built version of Sandcastle does include `require.js`, and I wasn't sure what the best approach here was. 

@emackey do you mind testing it one last time?

**Comment by OmarShehata** [comment]
@hpinkos @mramato do you think we can get this reviewed/merged in time for the next release?

**Comment by mramato** [comment]
Update `filesToClean` in gulpfile.js to include your new file.

**Comment by mramato** [comment]
This all sounds good to me.  Not sure about `bucketRaw.css` as the name, why not keep it `bucket.css` and have the new one be called `bucket.min.css` or similar?

@emackey can you do the final test/merge here?  I don't have access to IE11 at the moment and you know this code a lot better than me.  Thanks.

(And thanks again @OmarShehata sorry for the delay)

**Comment by OmarShehata** [comment]
@mramato thanks for the review! I'll address your suggestions. For the naming, I remember doing this because I thought it otherwise wouldn't be backwards compatible (since all the Sandcastles in the wild will try to import `bucket.css`) right?

**Comment by mramato** [comment]
That sounds like a good reason, but @emackey may know better.

**Comment by OmarShehata** [comment]
I made sure the generated bucket css gets removed on `npm run clean` and I made a final test that it still works after rebuilding, and on IE11. @emackey let me know if you notice anything else in your tests.

**Comment by emackey** [comment]
What happened with Travis CI?

> all the Sandcastles in the wild will try to import `bucket.css` right?

Correct.

Overall this looks good.  It no longer works in my IIS subfolder, but seems to work fine when used with `server.js` as intended.  Let's merge it once Travis is happy.

**Comment by mramato** [comment]
> Overall this looks good. It no longer works in my IIS subfolder, but seems to work fine when used with `server.js` as intended. Let's merge it once Travis is happy.

This doesn't sounds right.  Why doesn't it work under IIS?  This shouldn't have a server component at all since it needs to work out of S3 as well.

**Comment by OmarShehata** [comment]
Travis is probably failing due to https://github.com/AnalyticalGraphicsInc/cesium/issues/7249. I restarted the build.

It does work for me when served from a simple static server (with `python -m http.server`).

**Comment by emackey** [comment]
> This doesn't sounds right. Why doesn't it work under IIS?

There are absolute paths involved, or relative paths that go "too high".  I serve Cesium out of a `/Cesium` subfolder.  Using the python or node servers will serve it at the root level.

I spent a while this morning trying to narrow down exactly where the problem is.  It affects the "open in new window" feature of both built and non-built Sandcastle, with different errors for each.

**Comment by mramato** [comment]
That definitely sounds like something we should fix before merging this, because it means our deployed branches won't work either.

**Comment by emackey** [comment]
@OmarShehata Try launching your python server in the parent folder, above the Cesium source tree.

File: gulpfile.js [hunk 4]
```diff
@@ -1228,7 +1248,13 @@ function buildSandcastle() {
         })
         .pipe(gulp.dest('Build/Apps/Sandcastle'));
 
-    return streamToPromise(mergeStream(appStream, imageStream));
+    var standaloneStream = gulp.src([
+        'Apps/Sandcastle/standalone.html'
+        ])
+        .pipe(gulpReplace('../../ThirdParty/requirejs-2.1.20/require.js', '/Build/CesiumUnminified/Cesium.js'))
```
Comment (emackey) [inline gulpfile.js]: Here's one suspect path.  It looks server-root-relative.

Comment (OmarShehata) [reply gulpfile.js]: Okay yeah, I can recreated it locally now.

**Comment by OmarShehata** [comment]
Ok so this was the issue:

> Does it matter that in the built version I do /Build/CesiumUnminified/Cesium.js instead of ../../CesiumUnminified/Cesium.js ? I think the issue there is that the document head <base> is added after that loads, so while ../../CesiumUnminified/Cesium.js is the correct URL, subsequent requests are incorrect. I'm not sure if there's an easy way around that.

from https://github.com/AnalyticalGraphicsInc/cesium/pull/7250#issuecomment-440411049

This wasn't an issue for regular Sandcastle because it was still using `requirejs` in the built version (it's including _some_ modules straight out of `Source/`). I thought The Right Way was to avoid `require` in the built, but re-doing it this way, and adding the base href early on, solves this issue and makes everything a little simpler. It's now a little closer to the way `CesiumSandcastle.js` works.

I've tested it on every combination of browser (IE11, Firefox and Chrome), built/unbuilt, hosted out of `npm run start` and in a subdirectory with `python -m http.server`. @emackey I am confident this should work (except for too long urls on IE11 issue). Please let me know if you notice anything else.

**Comment by OmarShehata** [comment]
Sandcastle itself hasn't changed. It uses `Source/Cesium` in unbuilt and `CesiumUnminified` in built. Standalone mode now works this way as well. I confirmed this looking at the network tab as well.

What I meant by:

> This wasn't an issue for regular Sandcastle because it was still using requirejs in the built version (it's including some modules straight out of Source/).

Was that even though built Sandcastle uses the built version of Cesium, it still includes some modules from `Source/` (like `defined`). This is the current behavior in master/in production.

**Comment by emackey** [comment]
This works for me in IIS 👍

**Comment by mramato** [comment]
One last thing, update CHANGES

**Comment by OmarShehata** [comment]
@mramato I thought we don't put Sandcastle changes in CHANGES (https://github.com/AnalyticalGraphicsInc/cesium/pull/7152#discussion_r225603817). There's no change to the CesiumJS API here.

**Comment by mramato** [comment]
> @mramato I thought we don't put Sandcastle changes in CHANGES ([#7152 (comment)](https://github.com/AnalyticalGraphicsInc/cesium/pull/7152#discussion_r225603817)). There's no change to the CesiumJS API here.

That comment was regarding breaking changes, which for the most part don't apply to Sandcastle, we do mention new examples or functionality in Sandcastle when it happens.  In this case it also fixes a bug because Open in New Window used to not work in many cases and also couldn't allow you to share the url, right?

**Comment by OmarShehata** [comment]
Makes sense. Should be ready now!

**Pull request merged**